C++ compiler warning about size of `_jl_datatype_t`

I am seeing this warning when building Julia (OS X, clang 3.6.2):
```
./julia.h:289:5: warning: union has size 0 in C, non-zero size in C++
      [-Wextern-c-compat]
    union {
```
I see that this datatype is used from both C and C++ code.

One way to avoid this would be to add a `char dummy` to the union (and to update all routines that allocate memory for this type).

--- a/src/dump.c
+++ b/src/dump.c
@@ -527,7 +527,7 @@ static void jl_serialize_datatype(ios_t *s, jl_datatype_t *dt)
         write_int32(s, dt->alignment);
         write_int8(s, dt->haspadding);
         size_t fieldsize = jl_fielddesc_size(dt->fielddesc_type);
-        ios_write(s, (char*)&dt->fields32[0], nf * fieldsize);
+        ios_write(s, (char*)dt->fields, nf * fieldsize);
         jl_serialize_value(s, dt->types);
     }
 
@@ -1129,7 +1129,7 @@ static jl_value_t *jl_deserialize_datatype(ios_t *s, int pos, jl_value_t **loc)
         dt->alignment = read_int32(s);
         dt->haspadding = read_int8(s);
         size_t fieldsize = jl_fielddesc_size(fielddesc_type);
-        ios_read(s, (char*)&dt->fields32[0], nf * fieldsize);
+        ios_read(s, (char*)dt->fields, nf * fieldsize);
         dt->types = (jl_svec_t*)jl_deserialize_value(s, (jl_value_t**)&dt->types);
         jl_gc_wb(dt, dt->types);
     }

--- a/src/julia.h
+++ b/src/julia.h
@@ -290,11 +290,7 @@ typedef struct _jl_datatype_t {
     uint32_t uid;
     void *struct_decl;  //llvm::Value*
     void *ditype; // llvm::MDNode* to be used as llvm::DIType(ditype)
-    union {
-        jl_fielddesc8_t fields8[0];
-        jl_fielddesc16_t fields16[0];
-        jl_fielddesc32_t fields32[0];
-    };
+    size_t fields[];
 } jl_datatype_t;
 
 typedef struct {
@@ -718,26 +714,26 @@ STATIC_INLINE jl_value_t *jl_cellset(void *a, size_t i, void *x)
     static inline uint32_t jl_field_##f(jl_datatype_t *st, int i)       \
     {                                                                   \
         if (st->fielddesc_type == 0) {                                  \
-            return st->fields8[i].f;                                    \
+            return ((jl_fielddesc8_t*)st->fields)[i].f;                 \
         }                                                               \
         else if (st->fielddesc_type == 1) {                             \
-            return st->fields16[i].f;                                   \
+            return ((jl_fielddesc16_t*)st->fields)[i].f;                \
         }                                                               \
         else {                                                          \
-            return st->fields32[i].f;                                   \
+            return ((jl_fielddesc32_t*)st->fields)[i].f;                \
         }                                                               \
     }                                                                   \
     static inline void jl_field_set##f(jl_datatype_t *st, int i,        \
                                        uint32_t val)                    \
     {                                                                   \
         if (st->fielddesc_type == 0) {                                  \
-            st->fields8[i].f = val;                                     \
+            ((jl_fielddesc8_t*)st->fields)[i].f = val;                  \
         }                                                               \
         else if (st->fielddesc_type == 1) {                             \
-            st->fields16[i].f = val;                                    \
+            ((jl_fielddesc16_t*)st->fields)[i].f = val;                 \
         }                                                               \
         else {                                                          \
-            st->fields32[i].f = val;                                    \
+            ((jl_fielddesc32_t*)st->fields)[i].f = val;                 \
         }                                                               \
     }
 

